app/components/socket-provider.tsx

```diff
@@ -0,0 +1,61 @@
+"use client";
+
+import { createContext, useContext, useEffect, useState } from "react";
+import { io } from "socket.io-client";
+import type { SocketClient } from "@/types";
+
+const socket: SocketClient = io({ autoConnect: false });
+
+const SocketContext = createContext({
+  socket,
+  isDead: false,
+});
+
+export function useSocket() {
+  return useContext(SocketContext);
+}
```

Comment:
we should make this `null` by default and use the convention outlined in the React TypeScript Cheatsheet, that way we can ensure the socket is connected before accessing events with `useSocketEvent`
```suggestion
type SocketContextValue = {
  socket: SocketClient;
  isDead: boolean;
}

const SocketContext = createContext<SocketContextValue | null>(null);

export function useSocket() {
  const value = useContext(SocketContext);
  if (!value) {
    throw new Error("useSocket should be used in SocketProvider");
  }
  return value;
}
```

see: [React TypeScript Cheatsheet | Context | Without default context value](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/context#without-default-context-value)